### PR TITLE
Added linter and pre-commit hook

### LINKED ISSUE #15: Add linting package
With there being javascript code in here for the Builder class and (soon to be) tests; the linting package should be installed and a pre-commit hook added to make sure the code style is consistent.

### PR DESCRIPTION
**Pull request by matt-fidd**
Closes #15.

This PR is dependent on BritishYouthBandAssociation/eslint-config#17 and BritishYouthBandAssociation/eslint-config#19 being merged first.

### CODE REVIEW

File: build/Builder.js
Comment (rugulous) [inline build/Builder.js]: If we're doing tidyup these comments can go

File: build/Builder.js
Comment (rugulous) [inline build/Builder.js]: Assuming all the `const`s are after passing through the linter?

I'm not really a fan of putting the spaces into the array declaration to be honest, it feels weird?

File: build/test.js
Comment (rugulous) [inline build/test.js:5]: Guess this depends on your definition of a constant ;)

File: build/test.js
Comment (rugulous) [inline build/test.js]: ???

File: package.json
Comment (rugulous) [inline package.json]: There isn't an `index.js`... (is there?)

**Review by rugulous — CHANGES_REQUESTED** [review]

File: package.json
Comment (matt-fidd) [reply package.json]: Oops, you're right

File: build/Builder.js
Comment (matt-fidd) [reply build/Builder.js]: `const` in javascript refers to a variable that can not be explicitly defined after initialisation. It is normally best practice to initialise all variables that don't change as `const` to catch any errors if you might redefine something accidentally.

If we need to reassign a value we should use `let` but the default is assumed to be const.

Comment (matt-fidd) [reply build/Builder.js]: I can change the lint rule for that if you want. There was an object spacing rule I liked that looked like this:

```
const a = { b: 'c' };
--- instead of ---
const a = {b: 'c'};
```

I prefered the former so I matched the array bracket spacing rule, but if we want to change that then it's easily doable.

File: build/test.js
Comment (matt-fidd) [reply build/test.js:5]: Haha ;)

File: build/test.js
Comment (matt-fidd) [reply build/test.js]: There is a rule to standardise quote styles and I sort of abitrarily chose `'` over `"` because that's what the linting package was written in - if we want to change this then it's dead easy.

Comment (rugulous) [reply build/test.js]: I feel like this rule might become more frustrating than useful tbh... where does it stand on backticks?

File: build/Builder.js
Comment (rugulous) [reply build/Builder.js]: Perhaps `let` is just me being lazy then, as I'm never sure if I'm gonna write back to that var and it's a pain having to track it down and change it 😅

For me personally, I would prefer to define arrays as

```
const arr = [1, 2, 3]
```

and objects as multiline declarations

```
const obj = {
    A: 1,
    B: 7,
    C: "purple"
}
```

Comment (rugulous) [reply build/Builder.js:26]: Oh my word that obj was horrible, it was supposed to be one more indent in 😅

File: build/test.js
Comment (matt-fidd) [reply build/test.js]: It allows backticks as long as they're necessary, eg if there is a template literal in there. But if not it will recommend that you use a string instead.

The rule can be removed if needed - it was just sort of a skeleton of some rules I use; I'm sure we'll find ones we don't like and weed them out.

File: build/Builder.js
Comment (matt-fidd) [reply build/Builder.js]: That can certainly be arranged; it depends how opinionated we want to go.

I can scrap the arr rule (or invert it to require it to be build like your example?)
And [there is a rule](https://eslint.org/docs/rules/object-curly-newline) I can add to enforce objects to be declared as multilines?

File: build/test.js
Comment (rugulous) [reply build/test.js]: I'd say take this rule out for now; both `'` and `"` are equally valid quotes so limiting ourselves to one or the other doesn't seem optimal imo.

Fair enough if it was e.g. C# where one denotes a `string` and the other a `char`, but I don't see the benefit here

If it causes problems further down the line then we can absolutely look at pulling this back in

File: build/Builder.js
Comment (rugulous) [reply build/Builder.js]: Again I'm not convinced that this type of linting gives us any advantage? It might just be me used to not having a linter though...

I understand the purpose of a linter is to ensure code styles, but really if you look at `[1, 2, 3]` then you'll understand `[ 1, 2, 3 ]`?

For objects it does look cleaner split over multiple lines imo - you are more likely to notice an object declaration as 
```
const obj = {
    a: 1,
    b: 2,
    c: "purple"
};
```

than

```
const obj = { a:1, b:2, c: "purple" };
```

but again I'm not sure if going to that level of detail is really necessary **so long as the object is properly declared**

**Comment by rugulous** [comment]
This says it's still in draft, so I won't do any actions until you're happy to make it a "real" pull request :)

File: build/Builder.js
Comment (matt-fidd) [reply build/Builder.js]: In that case then, I will go through and refine the ruleset to best practices and error prevention as opposed to general code style - Thinking about it now I agree with you on that it doesn't really matter.

I'll keep this PR on hold then until I've got the new ruleset merged into BritishYouthBandAssociation/eslint-config and we'll revisit this.

File: build/test.js
Comment (matt-fidd) [reply build/test.js]: Agreed - as above I'll look into simplifying the rules down to what we really need then

**Comment by matt-fidd** [comment]
Update:

In light of recent discussions, this PR will stay in draft until a fix for BritishYouthBandAssociation/eslint-config#20 is merged.
This issue details that the rules should be simplified to enforce best practices and prevent errors further down the line, but should still allow personal choice.

**Review by rugulous — APPROVED** [review]

**Comment by rugulous** [comment]
@matt-fidd take this out of draft and should be good to go :)

**Pull request merged**
